### PR TITLE
Saml 1.1 Validate support

### LINKED ISSUE #1: Implement samlValidate endpoint
The `samlValidate` endpoint performs the same service ticket validation task as `serviceValidate`, but instead of accepting query parameters and returning custom XML, it accepts and returns SAML XML documents.

It is an optional part of the CAS 3.0 specification, but is required by some client implementations (e.g. [scm-cas-plugin](https://bitbucket.org/triologygmbh/scm-cas-plugin) for SCM Manager).

CAS Specification link: https://apereo.github.io/cas/5.0.x/protocol/CAS-Protocol-Specification.html#samlvalidate-cas-30

### PR DESCRIPTION
**Pull request by hamnis**
The reasoning for this commit is that this is well supported by CAS,
and it might be reasonable to include this in something that emulates
CAS.

I have also changed the LoginProtocol service to also allow login on the
root url. I can remove that if needed.

Closes #1

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -22,12 +22,12 @@
 
     <groupId>org.keycloak</groupId>
     <artifactId>keycloak-protocol-cas</artifactId>
-    <version>4.6.0</version>
+    <version>4.6.0-Get-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (hamnis) [reply pom.xml]: sure. this was just remnants of internal stuff.

File: src/main/java/org/keycloak/protocol/cas/CASLoginProtocol.java [hunk 2]
```diff
@@ -141,6 +142,11 @@ public Response finishLogout(UserSessionModel userSession) {
 
         event.event(EventType.LOGOUT);
         event.user(userSession.getUser()).session(userSession).success();
+
+        if (redirectUri != null) {
+            return Response.status(302).location(URI.create(redirectUri)).build();
+        }
```
[the remark this answers is not in the document]
Comment (hamnis) [reply src/main/java/org/keycloak/protocol/cas/CASLoginProtocol.java]: Will do

File: src/main/java/org/keycloak/protocol/cas/CASLoginProtocolService.java [hunk 3]
```diff
@@ -36,6 +33,11 @@ public static UriBuilder serviceBaseUrl(UriBuilder baseUriBuilder) {
         return baseUriBuilder.path(RealmsResource.class).path("{realm}/protocol/" + CASLoginProtocol.LOGIN_PROTOCOL);
     }
 
+    @Path("")
+    public Object root() {
+        return login();
+    }
```
[the remark this answers is not in the document]
Comment (hamnis) [reply src/main/java/org/keycloak/protocol/cas/CASLoginProtocolService.java]: Single client support, will remove

File: src/main/java/org/keycloak/protocol/cas/endpoints/ServiceValidateEndpoint.java [hunk 4]
```diff
@@ -26,6 +26,13 @@ public ServiceValidateEndpoint(RealmModel realm, EventBuilder event) {
 
     @Override
     protected Response successResponse() {
+        UserSessionModel userSession = clientSession.getUserSession();
+        Map<String, Object> attributes = getUserAttributes(clientSession, session);
+        CASServiceResponse serviceResponse = ServiceResponseHelper.createSuccess(userSession.getUser().getUsername(), attributes);
+        return prepare(Response.Status.OK, serviceResponse);
+    }
+
+    static Map<String, Object> getUserAttributes(AuthenticatedClientSessionModel clientSession, KeycloakSession session) {
```
[the remark this answers is not in the document]
Comment (hamnis) [reply src/main/java/org/keycloak/protocol/cas/endpoints/ServiceValidateEndpoint.java]: done

File: src/main/java/org/keycloak/protocol/cas/utils/AttributesMapAdapter.java [hunk 5]
```diff
@@ -17,7 +17,7 @@
 /**
  * Transforms the attribute map of the AuthenticationSuccess object (which can contain either simple values or
  * lists) to a flat list of XML nodes, where the key is the node name.<br>
- * Lists output multiple XML nodes with the same name.
+ * Lists soap multiple XML nodes with the same name.
```
[the remark this answers is not in the document]
Comment (hamnis) [reply src/main/java/org/keycloak/protocol/cas/utils/AttributesMapAdapter.java]: intellij being too eager with refactorings here

File: src/main/java/org/keycloak/protocol/cas/endpoints/SamlValidateEndpoint.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    @POST
+    @Consumes("text/xml;charset=utf-8")
+    @Produces("text/xml;charset=utf-8")
+    public Response validate(String input) {
+        MultivaluedMap<String, String> queryParams = request.getUri().getQueryParameters();
+        try {
+            String soapAction = Optional.ofNullable(request.getHttpHeaders().getHeaderString("SOAPAction")).map(s -> s.trim().replace("\"", "")).orElse("");
+            if (!soapAction.equals("http://www.oasis-open.org/committees/security")) {
+                throw new CASValidationException(CASErrorCode.INTERNAL_ERROR, "Not a validation request", Response.Status.BAD_REQUEST);
+            }
```
[the remark this answers is not in the document]
Comment (hamnis) [reply src/main/java/org/keycloak/protocol/cas/endpoints/SamlValidateEndpoint.java]: Required by saml 1.1 and the standard CAS library on the jvm sends that header.

**Comment by Doccrazy** [comment]
Thanks for the update, your changes are looking good! However, the compiled library is not working for me any more. When calling samlValidate I get:
```
Uncaught server error: java.lang.NoClassDefFoundError: org/keycloak/dom/saml/v1/protocol/SAML11ResponseType
        at org.keycloak.protocol.cas.representations.SamlResponseHelper.errorResponse(SamlResponseHelper.java:55)
        at org.keycloak.protocol.cas.endpoints.SamlValidateEndpoint.validate(SamlValidateEndpoint.java:67)
Caused by: java.lang.ClassNotFoundException: org.keycloak.dom.saml.v1.protocol.SAML11ResponseType from [Module "deployment.keycloak-protocol-cas-4.6.0.jar" from Service Module Loader]
        at org.jboss.modules.ModuleClassLoader.findClass(ModuleClassLoader.java:255)
        at org.jboss.modules.ConcurrentClassLoader.performLoadClassUnchecked(ConcurrentClassLoader.java:410)
        at org.jboss.modules.ConcurrentClassLoader.performLoadClass(ConcurrentClassLoader.java:398)
        at org.jboss.modules.ConcurrentClassLoader.loadClass(ConcurrentClassLoader.java:116)
```
I believe you need to define the new dependencies on the saml modules for OSGi in pom.xml (L130):
```xml
                        <manifestEntries>
                            <Dependencies>javax.xml.bind.api,org.keycloak.keycloak-core,org.keycloak.keycloak-server-spi,org.keycloak.keycloak-server-spi-private,org.keycloak.keycloak-services,org.keycloak.keycloak-saml-core,org.keycloak.keycloak-saml-core-public</Dependencies>
                        </manifestEntries>
```
Also, rebase please.

**Comment by hamnis** [comment]
Seems to be working with these latest changes.

**Comment by Doccrazy** [comment]
I tested your latest changes and everything appears to be working now, thanks!

**Comment by hamnis** [comment]
are you able to cut a new release of this?
Would also be grand if this was published to maven central or at least bintray.

**Comment by Doccrazy** [comment]
I don't really see the benefit in uploading to Maven Central. This library is not used as a dependency but has to be downloaded and installed manually. That is no more difficult from Github Releases than from Bintray. And the release cycle is not synchronized, so there's no guarantee that a new `keycloak-protcol-cas` will be available the moment Keycloak is pushed to central.  
But you could open an issue, and I or somebody else might look into it. 

I will upload a new release in the coming days.

**Comment by Doccrazy** [comment]
I published release 4.7.0 for the new Keycloak release. No changes required, so it should be backwards-compatible with 4.6.0. Hope this is ok for you, otherwise I would have to change the versioning scheme.

**Pull request merged**